model/resource/service_experimental.yaml

```diff
@@ -31,7 +31,28 @@ groups:
           the ID is ephemeral and changes during important lifetime events for the service
           (e.g. service restarts).
           If the service has no inherent unique ID that can be used as the value of this attribute
-          it is recommended to generate a random Version 1 or Version 4 RFC 4122 UUID
-          (services aiming for reproducible UUIDs may also use Version 5, see RFC 4122
-          for more recommendations).
+          it is recommended to generate a random Version 1 or Version 4
+          [RFC 4122](https://www.ietf.org/rfc/rfc4122.txt) UUID. Services aiming for reproducible UUIDs may also
+          use Version 5.
+
+          SDKs are required to follow the following algorithm when generating `service.instance.id`:
+
+          - If the user has provided a `service.instance.id`, via environment
+            variable, configuration or custom resource detection, this will
+            always be used and honored over generated IDs.
+          - When the `container.id` resource attribute is provided, it should be used as the input
+            for generating a UUID v5.
+          - When the SDK can detect it's running on Kubernetes or derivatives, such as by checking the
+            availability of a directory named `/var/run/secrets/kubernetes.io`, it should use
+            the contents of the file `/var/run/secrets/kubernetes.io/serviceaccount/namespace`,
+            along with the hostname and process ID (PID) separated with a dot ("namespace.hostname.pid") as the input
```

Comment:
I feel like this deserves an explanation: the PID here is to distinguish between containers within a pod. Different containers should yield different service.instance.id values.

Reply:
I like this. It services a similar function as `port` when doing prometheus metrics and identifying by the open port the container uses.

Reply:
Talking to @jsuereth via Slack made me realize that not all Kubernetes workload types should be treated equally: pods behind a stateful set or daemon set should strive to have the same instance id upon restarts, while pods for a deployment or batch jobs should not.

Reply:
Unfortunately, there doesn't seem to be a reliable way to get consistent IDs for stateful and daemon sets. Therefore, we'll rely on users to specify a custom attribute manually (`service.instance.id` itself or one that we can derive a consistent value, like `k8s.namespace.name`/`k8s.pod.name`/`k8s.container.name` for instance).